### PR TITLE
 Adding error variables in zeebe and connection timeout for api Paygops

### PR DESCRIPTION
**Pull request by ankita10r**
Capturing error and creating error codes and error related fields for Paygops confirmation and validation APIs.
This also has connection timeout for the same two api

@avikganguly01 and @fynmanoj : Please review

### CODE REVIEW

File: src/main/java/org/mifos/connector/ams/paygops/zeebe/ZeebeWorkers.java [hunk 1]
```diff
@@ -68,15 +69,24 @@ public void setupWorkers() {
                         producerTemplate.send("direct:transfer-validation-base", ex);
 
                         boolean isPartyLookUpFailed = ex.getProperty(PARTY_LOOKUP_FAILED, boolean.class);
-                        variables.put(PARTY_LOOKUP_FAILED, isPartyLookUpFailed);
+                        if(isPartyLookUpFailed) {
+                            variables.put(PARTY_LOOKUP_FAILED, true);
```
Comment (avikganguly01) [inline src/main/java/org/mifos/connector/ams/paygops/zeebe/ZeebeWorkers.java:73]: Line 73 to 76 is more or less duplicated everywhere. This should be separated as a function reused in all the 4 places.

Error information manually populated can be handled through control statements inside that function.

**Review by avikganguly01 — CHANGES_REQUESTED** [review]

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 2]
```diff
@@ -66,20 +66,31 @@ public void configure() {
...
+                        logger.info("body : "+ result);
+                        //JSONObject jsonObject = new JSONObject(body);
+                        if (result.getReconciled()){
                             logger.info("Paygops Validation Successful");
                             exchange.setProperty(PARTY_LOOKUP_FAILED, false);
                         }
                         else {
-                        logger.info("Paygops Validation Unsuccessful, Reconciled field returned false");
+                            logger.info("Paygops Validation Unsuccessful, Reconciled field returned false");
+                            exchange.setProperty(ERROR_DESCRIPTION, result);
```
Comment (avikganguly01) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: setCamelErrorInfo has to be made into a function and reused as well.

**Review by avikganguly01 — CHANGES_REQUESTED** [review]

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java (hunk 2 above)
Comment (ankita10r) [reply src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: okay

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 3]
```diff
@@ -66,20 +68,32 @@ public void configure() {
...
+                            String errorInfo = "Reconciled field returned false";
+                            String errorDesc = result.toString();
+                            setErrorCamelInfo(exchange,errorDesc,123,errorInfo);
+
+                            exchange.setProperty(PARTY_LOOKUP_FAILED, true);
                         }
-                        else {
-                        logger.info("Paygops Validation Unsuccessful, Reconciled field returned false");
+                    } catch (Exception e) {
+                        logger.info("Body data could not be parsed, setting validation as failed");
```
Comment (avikganguly01) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: Line 89,90,91 also goes inside the function to keep caller function clean.

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 4]
```diff
@@ -127,10 +140,19 @@ public void configure() {
...
-                    JSONObject jsonObject = new JSONObject(body);
-                    logger.info(jsonObject.toString());
-                    exchange.setProperty(TRANSFER_SETTLEMENT_FAILED, false);
+                    try {
+                        String body = exchange.getIn().getBody(String.class);
+                        JSONObject jsonObject = new JSONObject(body);
+                        exchange.setProperty(TRANSFER_SETTLEMENT_FAILED, false);
+                    } catch (Exception e) {
+                        logger.info("Body data could not be parsed, setting settlement as failed");
+                        String errorInfo = "Body data could not be parsed,setting confirmation as failed";
```
Comment (avikganguly01) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: Goes inside the function.

**Review by avikganguly01 — CHANGES_REQUESTED** [review]

**Comment by avikganguly01** [comment]
@ankita10r :Is there a similar PR for roster connector?

**Comment by ankita10r** [comment]
@avikganguly01 The roster workflow has a mechanism to do the same already

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java (hunk 3 above)
Comment (avikganguly01) [reply src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: @ankita10r : Did you get a chance to look at these review comments?

Comment (ankita10r) [reply src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: Sorry I had missed them, addressing them now

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 5]
```diff
@@ -66,20 +72,30 @@ public void configure() {
...
                     // processing success case
-                    String body = exchange.getIn().getBody(String.class);
-                    ObjectMapper mapper = new ObjectMapper();
-                    PaygopsResponseDto result = mapper.readValue(body, PaygopsResponseDto.class);
-                    logger.info("body : "+ result);
-                    //JSONObject jsonObject = new JSONObject(body);
-                    if (result.getReconciled()){
+                    try {
+                        String body = exchange.getIn().getBody(String.class);
+                        logger.info("raw reply : " + body);
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: donot use + with log ,  suggested change ` logger.info("raw reply : {}" , body);`

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 6]
```diff
@@ -66,20 +72,30 @@ public void configure() {
...
+                        ObjectMapper mapper = new ObjectMapper();
+                        PaygopsResponseDto result = mapper.readValue(body, PaygopsResponseDto.class);
+                        logger.info("after paygops dto : " + result);
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: donot use + with log

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java (hunk 5 above)
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: log level for this should not be info. for printing request response, recommended log level is trace

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 7]
```diff
@@ -66,20 +72,30 @@ public void configure() {
...
+                                    123,result.toString());
+
+                            exchange.setProperty(PARTY_LOOKUP_FAILED, true);
                         }
-                        else {
-                        logger.info("Paygops Validation Unsuccessful, Reconciled field returned false");
+                    } catch (Exception e) {
+                        logger.info(String.valueOf(e));
+                        setErrorCamelInfo(exchange,"Body data could not be parsed,setting validation as failed",
+                                123,exchange.getIn().getBody(String.class));
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: what is 123 here?

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 8]
```diff
@@ -88,10 +104,9 @@ public void configure() {
                     String body = exchange.getIn().getBody(String.class);
                     JSONObject jsonObject = new JSONObject(body);
                     Integer errorCode = jsonObject.getInt("error");
-                    String errorDescription = jsonObject.getString("error_message");
-                    exchange.setProperty(ERROR_CODE, errorCode);
-                    exchange.setProperty(ERROR_INFORMATION, jsonObject.toString(1));
-                    exchange.setProperty(ERROR_DESCRIPTION, errorDescription);
+                    String errorDescription   = jsonObject.getString("error_message");
+                    String errorInfo = jsonObject.toString(1);
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java:108]: we should have an error dictionary to mention all these error codes, we cannot randomly hardcode numbers like this. it should have a proper Enum

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 9]
```diff
@@ -115,7 +130,8 @@ public void configure() {
                     return verificationRequestDTO;
                 })
                 .marshal().json(JsonLibrary.Jackson)
-                .toD(getVerificationEndpoint() + "?bridgeEndpoint=true&throwExceptionOnFailure=false")
+                .toD(getVerificationEndpoint() + "?bridgeEndpoint=true&throwExceptionOnFailure=false&" +
+                        ConnectionUtils.getConnectionTimeoutDsl(amsTimeout))
                 .log(LoggingLevel.INFO, "Paygops validation api response: \n\n..\n\n..\n\n.. ${body}");
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: log lever : trace

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 10]
```diff
@@ -127,10 +143,17 @@ public void configure() {
...
-                    JSONObject jsonObject = new JSONObject(body);
-                    logger.info(jsonObject.toString());
-                    exchange.setProperty(TRANSFER_SETTLEMENT_FAILED, false);
+                    try {
+                        String body = exchange.getIn().getBody(String.class);
+                        JSONObject jsonObject = new JSONObject(body);
+                        exchange.setProperty(TRANSFER_SETTLEMENT_FAILED, false);
+                    } catch (Exception e) {
+                        setErrorCamelInfo(exchange,"Body data could not be parsed,setting confirmation as failed",
+                                123,exchange.getIn().getBody(String.class));
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java]: what is 123 ?

File: src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java [hunk 11]
```diff
@@ -140,9 +163,8 @@ public void configure() {
                     JSONObject jsonObject = new JSONObject(body);
                     Integer errorCode = jsonObject.getInt("error");
                     String errorDescription = jsonObject.getString("error_message");
-                    exchange.setProperty(ERROR_CODE, errorCode);
-                    exchange.setProperty(ERROR_INFORMATION, jsonObject.toString(1));
-                    exchange.setProperty(ERROR_DESCRIPTION, errorDescription);
+                    String errorInfo = jsonObject.toString(1);
```
Comment (fynmanoj) [inline src/main/java/org/mifos/connector/ams/paygops/camel/route/PaygopsRouteBuilder.java:165]: error code hardcoded

**Review by fynmanoj — CHANGES_REQUESTED** [review]

**Pull request merged**